### PR TITLE
Automated deployment using gh actions

### PR DESCRIPTION
**Pull request by vickywane**
This pull request contains the markdown file for my article on automating application deployments to Okteto Cloud using Okteto actions within a Github Actions Workflow. 

The article content in the `automated-deployments-using-github-actions.md` file follows the approved content outline in this [document](https://docs.google.com/document/d/18PjG2uOUDNF5MhjmbUNR4KvARyKWCd_ji8WLw01udRI/edit?usp=sharing). It is also available for review in a more readable format in this [dropbox paper](https://www.dropbox.com/scl/fi/7ovo5sopfqa0it0128tpm/Automating-Okteto-cloud-workflows-using-Okteto-Actions-and-GitHub-Actions..paper?dl=0&rlkey=mxsl7ie1dbikq8apv6406jju6).

CC: @rberrelleza @Youngestdev

### CODE REVIEW

**Comment by Youngestdev** [comment]
Hi @vickywane, I will be reviewing this blogpost. In the meantime, can you mail the images in a zip file to writers@okteto.com?

**Comment by vickywane** [comment]
Alright. I'll do that before the end of the day.

File: automated-deployments-using-github-actions.md [hunk 1]
```diff
@@ -0,0 +1,300 @@
+# Automating Okteto cloud workflows using Okteto Actions and GitHub Actions.
+
+## Introduction
+
+GitHub Actions provides developers with the functionality of building and automating  the execution of their software development workflows directly in a project’s repository.
+
+In this tutorial you will learn how you can automate the workflow of an existing Flask API whose continuous integration process is managed using [GitHub Actions](https://github.com/features/actions). You will learn about the Okteto actions and how you can use them to work with Okteto from a GitHub Action build.
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:7]: Hm, I'm not sure what you mean by Okteto actions. Do you mean to say GitHub actions created by Okteto?

File: automated-deployments-using-github-actions.md [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+For this article, we will be using [GitHub Actions](https://github.com/features/actions) as an automation tool to build a docker image from merged code changes and deploy to the docker image to the Okteto Cloud using the available [Okteto Actions](https://github.com/okteto/actions).
+
+## Prerequisites 
+
+In order to follow along with this tutorial, it is expected that you satisfy the following requirements;
+
+
+- Have an [Okteto Cloud](https://okteto.com/) Account
+- Have the [Okteto CLI](https://okteto.com/docs/getting-started/installation/index.html) installed on a host machine
+- Have a [GitHub account](https://github.com/) with Git installed on a host machine
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:25]: On your machine sounds better than 'on a host machine'.

File: automated-deployments-using-github-actions.md [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+- Have a basic understanding of the Python programming language
+- Have an installation of [Docker](https://www.docker.com/) on your local machine.
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:27]: Wouldn't "Have Docker installed on your machine" be better?

File: automated-deployments-using-github-actions.md [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+
+## Step 1: Clone a Sample Python Flask App
+
+To get started, fork the sample application from it’s repository [here](https://github.com/vickywane/okteto-flask-app). After forking the repository, execute the command below from your terminal to clone your forked copy of the repository to your host machine;
+
+**Note**: *Replace the* `*GITHUB_USERNAME*` *in the URL below with your GitHub username to match the remote origin of the forked repository.*
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:34]: Use a block quote.

File: automated-deployments-using-github-actions.md [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+- Have an installation of [Docker](https://www.docker.com/) on your local machine.
+
+
+## Step 1: Clone a Sample Python Flask App
+
+To get started, fork the sample application from it’s repository [here](https://github.com/vickywane/okteto-flask-app). After forking the repository, execute the command below from your terminal to clone your forked copy of the repository to your host machine;
+
+**Note**: *Replace the* `*GITHUB_USERNAME*` *in the URL below with your GitHub username to match the remote origin of the forked repository.* 
+
+
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:36]: Use ```console for command line code blocks.

File: automated-deployments-using-github-actions.md [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+    git clone https://github.com/{{GITHUB_USERNAME}}/okteto-flask-app
+
+ 
+Going through the `flaskr.py` file within the cloned `okteto-flask-app` folder in your preferred code editor, you would observe that it contains three API routes that are used for performing a create, retrieve and delete operation against a connected Couch database.
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:40]: "In the flaskr.py, there are three routes defined...".

File: automated-deployments-using-github-actions.md (hunk 4 above)
Comment (Youngestdev) [reply automated-deployments-using-github-actions.md:34]: Likewise, a $ before the command. E.g

```console
$ git clone xyz

File: automated-deployments-using-github-actions.md [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+
+The cloned project also contains a `Dockerfile` and `docker-compose.yml` file that contains all defined steps and services needed to create a multi-stage build for this application. 
+
+To run this application, execute the [docker-compose](https://docs.docker.com/compose/) command below to build and run the application container that consists of a `database` and `app` service.
+
+
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:46]: Add the back ticks.

File: automated-deployments-using-github-actions.md [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+    curl -X POST -d '{"name":"Victory Nwani","occupation":"Software Engineer"}'  -H 'Content-Type: application/json' http://localhost:5050/api/customer
+
+To view the data inserted from the `POST` request above, execute the command below from your terminal to make a GET request to the `/api/customer` api route within the flask API using [cURL](https://curl.se/) which retrieves all documents in the customer collection within the couch database.
+
+
+    curl http://localhost:5050/api/customer
+
+
+![](https://lh3.googleusercontent.com/1zojwJ4AEJ9siyHAOxv4Vexk6tK3-UsCW9CBB6KyZKihQUMebSi2xbk11VFPnBWvybnXXK3ZfgyjXD400w-Nzj2JNK3rA6Q08gx8bPQl75Igw7dVFQo7I7VHMPn0ml9j_7MmE7G_)
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:60]: Replace the links here with the name of the image from the zip file you sent. Also, add a text in the []. 


Apply this change uniformly.

File: automated-deployments-using-github-actions.md [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+
+## Step 2: Deploy Flask Application To Okteto Cloud 
+
+Now you have the cloned application working locally on your computer. You will deploy the cloned version to the Okteto cloud to serve as your production deployment. The folder already contains a `Dockerfile` and a `docker-compose.yml` file that defines the services and steps needed to build the image for this application. 
+
+To begin the deployment from your terminal using the Okteto CLI, execute the command below to login and create a session between your okteto cloud account and your local terminal;
+
+
+    okteto login
+
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:76]: Add:

```console
$ okteto namespace
```

Explain briefly what this command above does too.

File: automated-deployments-using-github-actions.md [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+           python-version: 3
+    
+       - name: Install python dependencies
+         run: pip install -r requirements.txt
+
+Next, to begin automating the deployment process of the application to okteto, use the [okteto-login](https://github.com/okteto/login) action to create a session between the workflow build and your namespace. 
+
+Add the code block below into the `ci.yml` file. The code block contains a new step that uses the [okteto-login](https://github.com/okteto/login) action with your okteto [personal access token](https://okteto.com/docs/cloud/personal-access-tokens/index.html) in the step’s environment variables to create a session between your okteto account and the current workflow build.
+
+**Note**: See the [Personal Access Token](https://okteto.com/docs/cloud/personal-access-tokens/index.html) section of the Okteto developer documentation on how to create your a token from the your Okteto cloud account or Okteto CLI.
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:194]: Block quote here.

File: automated-deployments-using-github-actions.md [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+
+Commit and push the changes to the `ci.yml` file above to trigger the actions check for the pull request.
+
+Going through the workflow logs shown in the image below, you would notice that the highlighted “**Build and deploy application image to Production Okteto Namespace**” step in the build workflow was not executed as expected. 
+
+
+![](https://lh6.googleusercontent.com/UBMoSEGmy3QvGd3mLRf7JdiYBqHlcLQkGRgc5oA7w1XVyRspc84UKSNoIFLNNc-qTz0lqxxTe12_OK7MPEQ4BcW1MgeW1Aa-enB7PL6nVW79QO_ztfZnpbT3toLzmWM1MAAn8qE0)
+
+
+**Note**: When working with pull requests for your project managed with Okteto, you should set up a [preview deployment](https://okteto.com/blog/preview-environments-for-kubernetes/) workflow. [This](https://okteto.com/blog/deploying-preview-environments-for-docker-compose-applications-using-okteto-and-github/) tutorial explains the process of creating a preview deployment workflow for Docker compose applications.
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:278]: Block quote here.

File: automated-deployments-using-github-actions.md [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+![](https://lh6.googleusercontent.com/4zdWAWhc01t4zJHQKEzj669xagWIpJqQsW74mL5vZZK-wzlVaW96z7YPGtbufyiI7aQY26eteB1p55ldgNRwVzOc7r1imZO2L8VB1gTRr9qU8ZC0irR4Q12cManc9MqV10V1dQ7l)
+
+
+From the image above, you can observe that the build was successful and it was triggered by a push event to the master branch which was automatically triggered by GitHub when you merged the pull request. This push event would further cause the if condition in the “Build and deploy” step to evaluate to `true` and execute the step.
+
+All subsequent commits and pull requests would go through this workflow build process, automatically deploying new changes to your Okteto namespace after new changes have been merged into your master branch.
+
+# Conclusion 
+
+Within this article, you cloned a sample python application, built and deployed a docker image of the application to the Okteto cloud, then you automated the testing and deployment of your application’s docker  image to your cloud Okteto namespace using a GitHub Action workflow.
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:294]: In this article^.

**Review by Youngestdev — COMMENTED** [review]
- Use block quotes for notes. 
- Rename image paths to the name of the image in the zip files. 
- Add the back ticks indicating a code block 
- Check other comments I've made. 

Cheers!

File: automated-deployments-using-github-actions.md (hunk 1 above)
Comment (vickywane) [reply automated-deployments-using-github-actions.md:7]: Alright.

File: automated-deployments-using-github-actions.md (hunk 2 above)
Comment (vickywane) [reply automated-deployments-using-github-actions.md:25]: Alright. I would rephrase this.

I however know of individuals who try out tutorials on a VM being a 'host machine'

File: automated-deployments-using-github-actions.md (hunk 3 above)
Comment (vickywane) [reply automated-deployments-using-github-actions.md:27]: Yes, that would make it uniform with the previous corrections.

File: automated-deployments-using-github-actions.md (hunk 4 above)
Comment (vickywane) [reply automated-deployments-using-github-actions.md:34]: Alright. I would make this adjustment

File: automated-deployments-using-github-actions.md (hunk 6 above)
Comment (vickywane) [reply automated-deployments-using-github-actions.md:40]: Alright. I have rephrased the sentence into this.

**Comment by vickywane** [comment]
I have just applied your comments within this pull request to the file. 

I however would need you to clarify how cingle line commands should be written. 
Should they be written with an indented space ( the markdown command syntax )?
    
      npm test
 
Or should the command be surrounded with backticks as shown below? 

`npm test`

File: automated-deployments-using-github-actions.md [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+All subsequent commits and pull requests would go through this workflow build process, automatically deploying new changes to your Okteto namespace after new changes have been merged into your master branch.
+
+# Conclusion
+
+Within this article, you cloned a sample python application, built and deployed a docker image of the application to the Okteto cloud, then you automated the testing and deployment of your application’s docker  image to your cloud Okteto namespace using a GitHub Action workflow.
+
+As suggested within the article, you can set up a [deployment preview](https://okteto.com/blog/preview-environments-for-kubernetes/) workflow to create a preview separately for each pull request opened in your project’s repository. [This](https://okteto.com/blog/deploying-preview-environments-for-docker-compose-applications-using-okteto-and-github/) tutorial explains the process of creating a preview deployment workflow for Docker compose applications.
+
+The completed application is available on GitHub and can be found in [this repository](https://github.com/vickywane/okteto-flask-app), for you to clone and use.
+``````````
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:334]:
```suggestion
```

File: automated-deployments-using-github-actions.md [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+    
+       - name: Install python dependencies
+         run: pip install -r requirements.txt
+```
+
+Next, to begin automating the deployment process of the application to okteto, use the [okteto-login](https://github.com/okteto/login) action to create a session between the workflow build and your namespace.
+
+Add the code block below into the `ci.yml` file. The code block contains a new step that uses the [okteto-login](https://github.com/okteto/login) action with your okteto [personal access token](https://okteto.com/docs/cloud/personal-access-tokens/index.html) in the step’s environment variables to create a session between your okteto account, and the current workflow build.
+
+> **Note**: See the [Personal Access Token](https://okteto.com/docs/cloud/personal-access-tokens/index.html) section of the Okteto developer documentation on how to create a token using the Okteto cloud or Okteto CLI.
```
[the remark this answers is not in the document]
Comment (vickywane) [reply automated-deployments-using-github-actions.md:227]: Would paragraph still be neccessary considering that I would add instructions on how to create the access token?

File: automated-deployments-using-github-actions.md [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+
+Next, verify your Okteto namespace is active using the [okteto-namespace](https://github.com/okteto/namespace) action within the step in the code block below;
+
+
+```yaml
+- name: Verify okteto namespace
+  uses: okteto/namespace@master
+  with:
+    namespace: vickywane
+```
```
[the remark this answers is not in the document]
Comment (vickywane) [reply automated-deployments-using-github-actions.md:245]: Alright.

File: automated-deployments-using-github-actions.md [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+         with:
+           python-version: 3
+    
+       - name: Install python dependencies
+         run: pip install -r requirements.txt
+```
+
+Next, to begin automating the deployment process of the application to okteto, use the [okteto-login](https://github.com/okteto/login) action to create a session between the workflow build and your namespace.
+
+Add the code block below into the `ci.yml` file. The code block contains a new step that uses the [okteto-login](https://github.com/okteto/login) action with your okteto [personal access token](https://okteto.com/docs/cloud/personal-access-tokens/index.html) in the step’s environment variables to create a session between your okteto account, and the current workflow build.
```
[the remark this answers is not in the document]
Comment (vickywane) [reply automated-deployments-using-github-actions.md:225]: Alright. 

Initially skipped it and linked the official docs. I'll add it, and update the initial ZIP file

**Comment by vickywane** [comment]
Hi, @Youngestdev,

 I have applied your suggestions and also made the sited adjustments. 
I would update the ZIP file containing the images and resend it to you.

File: automated-deployments-using-github-actions.md (hunk 9 above)
Comment (Youngestdev) [reply automated-deployments-using-github-actions.md:82]: @vickywane.

**Comment by Youngestdev** [comment]
Thank you for the article draft and sorry for the long wait too. @rberrelleza will be making a final review this week :-)

**Review by Youngestdev — APPROVED** [review]
There's a bit of formatting issues but I'll take care of that.

**Comment by vickywane** [comment]
Thanks @Youngestdev !

Looking forward to @rberrelleza review of the draft.

File: automated-deployments-using-github-actions.md [hunk 17]
```diff
@@ -0,0 +1,400 @@
...
+      uses: okteto/login@master
+      with:
+        token: ${{ secrets.OKTETO_TOKEN }}
+
+    - name: Activate application namespace
+      uses: okteto/namespace@master
+      with:
+        namespace: vickywane
+
+    - name: Build and deploy application image to Production Okteto Namespace
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:364]:
```suggestion
    - name: Build and deploy application image to Okteto
```

File: automated-deployments-using-github-actions.md [hunk 18]
```diff
@@ -0,0 +1,400 @@
...
+
+
+You can also work with the running couch database through the [Fauxton web interface](https://couchdb.apache.org/fauxton-visual-guide/index.html) that comes by default with couchdb image at [http://localhost:5984/_utils](http://localhost:5984/_utils).
+
+> As defined in the environment field within the database service in the `docker-compose.yml` file, the couchdb username is **couchdb-admin**, while the password is **couchdb-password**. Feel free to change them to your own preferred secured values.
+
+
+## Step 2: Deploy Flask Application To Okteto Cloud
+
+Now you have the cloned application working locally on your computer. You will deploy the cloned version to the Okteto cloud to serve as your production deployment. The folder already contains a `Dockerfile` and a `docker-compose.yml` file that defines the services and steps needed to build the image for this application.
```
Comment (Youngestdev) [inline automated-deployments-using-github-actions.md:75]: So, Okteto isn't for production based deployment. Okteto is primarily for development so the theme for this article should be for development.

**Review by Youngestdev — CHANGES_REQUESTED** [review]

**Comment by Youngestdev** [comment]
Some comments @vickywane from Ramiro:

1. A good scenario would be to deploy during merge to validate that the changes don’t break, and then many run an end to end test to validate that the app works.

File: automated-deployments-using-github-actions.md (hunk 18 above)
Comment (vickywane) [reply automated-deployments-using-github-actions.md:75]: Alright. 

I have applied this correction and adjusted the article's title.

**Comment by vickywane** [comment]
> Some comments @vickywane from Ramiro:
>
>    A good scenario would be to deploy during merge to validate that the changes don’t break, and then many run an end to end > test to validate that the app works.

@Youngestdev The comment here is quite unclear as i think it has already been covered in the article draft.

- Unit tests were written for the API endpoints using `httppretty` and `pytest` in **step 3**.
- The `.Dockerfile` included a command to execute the unit test in `line 191` of step 3.
- A merge of the pull request opened within the article was done and the tests were executed directly agaisnt the master branch ( including a deployment to the Okteto namespace ) in **step 5**.

Did @rberrelleza look through the images? ( They were shown in the images )

**Comment by Youngestdev** [comment]
Hi @vickywane,

Ramiro requested a further review: *so for this post, I was thinking about, what if you have a test that you run everything you merge to main? Okteto can be used to deploy an ‘on demand’ copy of the app with the latest code, and then you can run a test to make sure that everything works fine, by calling the application directly, instead of using mocks.*

So our CI has two jobs:

- Build a copy of the application first
- Run the tests on the endpoint of the copy of the application built

**Review by Youngestdev — APPROVED** [review]
